librz/analysis/p/analysis_evm_cs.c

```diff
@@ -0,0 +1,419 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_analysis.h>
+#include <rz_lib.h>
+#include <capstone/capstone.h>
+#include <capstone/sparc.h>
+
+#if CS_API_MAJOR < 5
+#error Old Capstone not supported
+#endif
+
+#define INSOP(n) insn->detail->sparc.operands[n]
+#define INSCC    insn->detail->sparc.cc
+
+static void opex(RzStrBuf *buf, csh handle, cs_insn *insn) {
+	int i;
+	PJ *pj = pj_new();
+	if (!pj) {
+		return;
+	}
+	pj_o(pj);
+	pj_ka(pj, "operands");
+	cs_sparc *x = &insn->detail->sparc;
+	for (i = 0; i < x->op_count; i++) {
+		cs_sparc_op *op = x->operands + i;
+		pj_o(pj);
+		switch (op->type) {
+		case SPARC_OP_REG:
+			pj_ks(pj, "type", "reg");
+			pj_ks(pj, "value", cs_reg_name(handle, op->reg));
+			break;
+		case SPARC_OP_IMM:
+			pj_ks(pj, "type", "imm");
+			pj_kN(pj, "value", op->imm);
+			break;
+		case SPARC_OP_MEM:
+			pj_ks(pj, "type", "mem");
+			if (op->mem.base != SPARC_REG_INVALID) {
+				pj_ks(pj, "base", cs_reg_name(handle, op->mem.base));
+			}
+			pj_ki(pj, "disp", op->mem.disp);
+			break;
+		default:
+			pj_ks(pj, "type", "invalid");
+			break;
+		}
+		pj_end(pj); /* o operand */
+	}
+	pj_end(pj); /* a operands */
+	pj_end(pj);
+
+	rz_strbuf_init(buf);
+	rz_strbuf_append(buf, pj_string(pj));
+	pj_free(pj);
+}
+
+static int parse_reg_name(RzRegItem *reg, csh handle, cs_insn *insn, int reg_num) {
+	if (!reg) {
+		return -1;
+	}
+	switch (INSOP(reg_num).type) {
+	case SPARC_OP_REG:
+		reg->name = (char *)cs_reg_name(handle, INSOP(reg_num).reg);
+		break;
+	case SPARC_OP_MEM:
+		if (INSOP(reg_num).mem.base != SPARC_REG_INVALID) {
+			reg->name = (char *)cs_reg_name(handle, INSOP(reg_num).mem.base);
+			break;
+		}
+	default:
+		break;
+	}
+	return 0;
+}
+
+static void op_fillval(RzAnalysisOp *op, csh handle, cs_insn *insn) {
+	static RzRegItem reg;
+	switch (op->type & RZ_ANALYSIS_OP_TYPE_MASK) {
+	case RZ_ANALYSIS_OP_TYPE_LOAD:
+		if (INSOP(0).type == SPARC_OP_MEM) {
+			ZERO_FILL(reg);
+			op->src[0] = rz_analysis_value_new();
+			op->src[0]->type = RZ_ANALYSIS_VAL_MEM;
+			op->src[0]->reg = &reg;
+			parse_reg_name(op->src[0]->reg, handle, insn, 0);
+			op->src[0]->delta = INSOP(0).mem.disp;
+		}
+		break;
+	case RZ_ANALYSIS_OP_TYPE_STORE:
+		if (INSOP(1).type == SPARC_OP_MEM) {
+			ZERO_FILL(reg);
+			op->dst = rz_analysis_value_new();
+			op->dst->type = RZ_ANALYSIS_VAL_MEM;
+			op->dst->reg = &reg;
+			parse_reg_name(op->dst->reg, handle, insn, 1);
+			op->dst->delta = INSOP(1).mem.disp;
+		}
+		break;
+	}
+}
+
+static int analop(RzAnalysis *a, RzAnalysisOp *op, ut64 addr, const ut8 *buf, int len, RzAnalysisOpMask mask) {
+	static csh handle = 0;
+	static int omode;
+	cs_insn *insn;
+	int mode, n, ret;
+
+	if (!a->big_endian) {
+		return -1;
+	}
+
+	mode = CS_MODE_LITTLE_ENDIAN;
+	if (!strcmp(a->cpu, "v9")) {
+		mode |= CS_MODE_V9;
+	}
+	if (mode != omode) {
+		cs_close(&handle);
+		handle = 0;
+		omode = mode;
+	}
+	if (handle == 0) {
+		ret = cs_open(CS_ARCH_SPARC, mode, &handle);
```

Comment:
Defininetly not `SPARC`

Reply:
The first goal is to just register that arch and anal plugins in the core. The whole logic is completely untouched yet.